File: src/AeterniUI/Components/FlashCard/FlashCard.razor.js
// FlashCard pointer following. The module only mirrors the pointer position into
// two rotation custom properties and the sheen position on the card root, and
// toggles `is-tilting`; the rotation, the hover scale, the flip, the sheen finish
// and every transition stay in the isolated stylesheet. The flipped state belongs
// to the component, not to the browser.
//
// The effect is skipped for `Tilt="false"`, for disabled cards and for the
// `prefers-reduced-motion: reduce` preference. Those cards keep the flip, the
// focus ring and the resting sheen, and simply never tilt or track the pointer.
//
// The pointer rectangle is cached on `pointerenter` instead of being measured on
// every move: a tilted card reports an inflated bounding box, which would feed
// the rotation back into itself and make the card jitter.

const instances = new Map();

export function attach(instanceId, root, enabled, maxAngle, sheen) {
    detach(instanceId);

    if (!instanceId || !root) {
        return;
    }

    const state = {
        root,
        angle: Number(maxAngle) || 0,
        sheen: !!sheen,
        rect: null,
        onEnter: null,
        onMove: null,
        onLeave: null
    };

    instances.set(instanceId, state);

    if (!enabled || state.angle <= 0 || prefersReducedMotion()) {
        return;
    }

    state.onEnter = () => { state.rect = root.getBoundingClientRect(); };
    state.onMove = event => track(state, event);
    state.onLeave = () => reset(state);

    root.addEventListener('pointerenter', state.onEnter);
    root.addEventListener('pointermove', state.onMove);
    root.addEventListener('pointerleave', state.onLeave);
    root.addEventListener('pointercancel', state.onLeave);
}

export function dispose(instanceId) {
    const state = instances.get(instanceId);
    if (!state) {
        return;
    }

    release(state);
    instances.delete(instanceId);
}

function detach(instanceId) {
    if (instanceId && instances.has(instanceId)) {
        dispose(instanceId);
    }
}

// Every render re-attaches the instance, so releasing the listeners and clearing
// the transform has to stay in one place: a detached card must not keep a stale
// rotation when its instance is replaced.
function release(state) {
    if (state.onEnter) {
        state.root.removeEventListener('pointerenter', state.onEnter);
    }
    if (state.onMove) {
        state.root.removeEventListener('pointermove', state.onMove);
    }
    if (state.onLeave) {
        state.root.removeEventListener('pointerleave', state.onLeave);
        state.root.removeEventListener('pointercancel', state.onLeave);
    }

    state.onEnter = null;
    state.onMove = null;
    state.onLeave = null;
    reset(state);
}

function track(state, event) {
    const rect = state.rect ?? state.root.getBoundingClientRect();
    if (!rect || !rect.width || !rect.height) {
        return;
    }

    const x = clamp((event.clientX - rect.left) / rect.width - 0.5);
    const y = clamp((event.clientY - rect.top) / rect.height - 0.5);

    // The side under the pointer leans towards the viewer, capped at ±MaxTiltAngle
    // on the edges.
    state.root.style.setProperty('--aeterni-flash-card-rotate-y', `${(-x * 2 * state.angle).toFixed(2)}deg`);
    state.root.style.setProperty('--aeterni-flash-card-rotate-x', `${(y * 2 * state.angle).toFixed(2)}deg`);

    if (state.sheen) {
        state.root.style.setProperty('--aeterni-flash-card-sheen-x', `${((x + 0.5) * 100).toFixed(1)}%`);
        state.root.style.setProperty('--aeterni-flash-card-sheen-y', `${((y + 0.5) * 100).toFixed(1)}%`);
    }

    state.root.classList.add('is-tilting');
}

function reset(state) {
    state.rect = null;
    state.root.classList.remove('is-tilting');
    state.root.style.setProperty('--aeterni-flash-card-rotate-x', '0deg');
    state.root.style.setProperty('--aeterni-flash-card-rotate-y', '0deg');
}

function clamp(value) {
    return Math.max(-0.5, Math.min(0.5, value));
}

function prefersReducedMotion() {
    return typeof window !== 'undefined'
        && typeof window.matchMedia === 'function'
        && window.matchMedia('(prefers-reduced-motion: reduce)').matches;
}

File: src/AeterniUI/Components/FlashCard/FlashCard.razor.cs
using System.Globalization;
using AeterniUI.Attributes;
using AeterniUI.Enums;
using Microsoft.AspNetCore.Components;
using Microsoft.AspNetCore.Components.Web;

namespace AeterniUI.Components.FlashCard;

/// <summary>
/// An image flash card: a front face with media and an optional caption, plus an
/// optional back face that the user reveals by flipping the card. The pointer
/// following 3D tilt is the only browser behaviour; the flip is component state,
/// so the rendered sides and their ARIA stay in sync with <see cref="IsFlipped"/>.
/// </summary>
[JsModule("Components/FlashCard/FlashCard.razor.js", Name = "flash-card")]
public partial class FlashCard : AeterniComponent
{
    private const string ModuleName = "flash-card";

    private bool _internalFlipped;
    private bool _stateInitialized;

    /// <summary>Image shown on the front face.</summary>
    [Parameter]
    public string? ImageSrc { get; set; }

    /// <summary>
    /// Alternative text for <see cref="ImageSrc" />. An empty value marks the image
    /// as decorative; supply it whenever the image carries the card's meaning.
    /// </summary>
    [Parameter]
    public string? ImageAlt { get; set; }

    /// <summary>Width to height ratio of the card face.</summary>
    [Parameter]
    public double AspectRatio { get; set; } = 1;

    /// <summary>Front face heading.</summary>
    [Parameter]
    public string? Title { get; set; }

    /// <summary>Front face supporting text.</summary>
    [Parameter]
    public string? Description { get; set; }

    /// <summary>Back face content. Supplying it makes the card flippable.</summary>
    [Parameter]
    public RenderFragment? Back { get; set; }

    /// <summary>Whether the back face is currently shown.</summary>
    [Parameter]
    public bool IsFlipped { get; set; }

    /// <summary>Raised before the card is flipped, with the requested state.</summary>
    [Parameter]
    public EventCallback<bool> IsFlippedChanged { get; set; }

    /// <summary>
    /// How the back face is revealed. <see cref="FlashCardFlipTrigger.Click" /> is
    /// the default keyboard reachable control; <see cref="FlashCardFlipTrigger.Hover" />
    /// is a pointer-only preview that never changes <see cref="IsFlipped" />.
    /// </summary>
    [Parameter]
    public FlashCardFlipTrigger FlipTrigger { get; set; } = FlashCardFlipTrigger.Click;

    /// <summary>Whether the pointer follows the tilt. Reduced motion always disables it.</summary>
    [Parameter]
    public bool Tilt { get; set; } = true;

    /// <summary>Maximum rotation in degrees reached at the card edges.</summary>
    [Parameter]
    public double MaxTiltAngle { get; set; } = 16;

    /// <summary>Scale applied while the pointer is over the card.</summary>
    [Parameter]
    public double TiltScale { get; set; } = 1.02;

    /// <summary>CSS perspective distance in pixels.</summary>
    [Parameter]
    public double Perspective { get; set; } = 800;

    /// <summary>
    /// Surface finish of the media area. <see cref="FlashCardSheen.Holo" /> is the
    /// rainbow diffraction finish of a holographic trading card,
    /// <see cref="FlashCardSheen.Shine" /> the metallic white sweep, and
    /// <see cref="FlashCardSheen.None" /> renders no overlay.
    /// </summary>
    [Parameter]
    public FlashCardSheen Sheen { get; set; } = FlashCardSheen.Holo;

    /// <summary>Strength of the sheen overlay, between 0 and 1.</summary>
    [Parameter]
    public double SheenIntensity { get; set; } = 0.6;

    /// <summary>Accessible name of the flip button; falls back to <see cref="Title" />.</summary>
    [Parameter]
    public string? AriaLabel { get; set; }

    protected override void OnParametersSet()
    {
        base.OnParametersSet();

        if (!_stateInitialized && !IsControlled)
        {
            _internalFlipped = IsFlipped;
            _stateInitialized = true;
        }

        if (!double.IsFinite(AspectRatio) || AspectRatio <= 0)
        {
            throw new ArgumentOutOfRangeException(nameof(AspectRatio), AspectRatio, "AspectRatio must be a positive finite number.");
        }

        if (!Enum.IsDefined(FlipTrigger))
        {
            throw new ArgumentOutOfRangeException(nameof(FlipTrigger), FlipTrigger, "Unknown flash card flip trigger.");
        }

        if (!double.IsFinite(Perspective) || Perspective <= 0)
        {
            throw new ArgumentOutOfRangeException(nameof(Perspective), Perspective, "Perspective must be a positive finite number.");
        }

        if (!double.IsFinite(MaxTiltAngle) || MaxTiltAngle < 0)
        {
            throw new ArgumentOutOfRangeException(nameof(MaxTiltAngle), MaxTiltAngle, "MaxTiltAngle must be a non-negative finite number.");
        }

        if (!double.IsFinite(TiltScale) || TiltScale <= 0)
        {
            throw new ArgumentOutOfRangeException(nameof(TiltScale), TiltScale, "TiltScale must be a positive finite number.");
        }

        if (!Enum.IsDefined(Sheen))
        {
            throw new ArgumentOutOfRangeException(nameof(Sheen), Sheen, "Unknown flash card sheen finish.");
        }

        if (!double.IsFinite(SheenIntensity) || SheenIntensity < 0 || SheenIntensity > 1)
        {
            throw new ArgumentOutOfRangeException(nameof(SheenIntensity), SheenIntensity, "SheenIntensity must be between 0 and 1.");
        }

        if (!HasImage && !HasTitle && !HasDescription && Back is null)
        {
            throw new ArgumentException("FlashCard requires ImageSrc, Title, Description, or Back content.");
        }
    }

    protected override ClassBuilder BuildClass() => base.BuildClass()
        .Add("aeterni-flash-card")
        .Add(ComponentClass.For("aeterni-flash-card", SheenClass))
        .Add("is-flippable", CanFlip)
        .Add("is-flip-hover", HoverPreview)
        .Add("is-flipped", CanFlip && CurrentFlipped)
        .Add("is-disabled", Disabled);

    // The pointer rotation, the hover scale and the sheen position are consumed by
    // the isolated stylesheet; the angle itself is computed in JS from MaxTiltAngle.
    protected override StyleBuilder BuildStyle() => base.BuildStyle()
        .Add("--aeterni-flash-card-aspect", AspectRatio.ToString(CultureInfo.InvariantCulture))
        .Add("--aeterni-flash-card-scale", TiltScale.ToString(CultureInfo.InvariantCulture))
        .Add("--aeterni-flash-card-sheen-opacity", SheenIntensity.ToString(CultureInfo.InvariantCulture))
        .Add("--aeterni-flash-card-perspective", $"{Perspective.ToString(CultureInfo.InvariantCulture)}px");

    protected override IReadOnlyDictionary<string, object> BuildAttributes()
    {
        var attributes = new Dictionary<string, object>(base.BuildAttributes(), StringComparer.OrdinalIgnoreCase);

        if (!UsesClickFlip)
        {
            return attributes;
        }

        attributes["role"] = "button";
        attributes["aria-pressed"] = CurrentFlipped ? "true" : "false";

        if (Disabled)
        {
            // The root is a div, so it cannot carry the native disabled attribute:
            // the state is announced and the card leaves the tab sequence instead.
            attributes["aria-disabled"] = "true";
        }
        else
        {
            attributes["tabindex"] = "0";
        }

        var label = FirstNonBlank(AriaLabel, Title);
        if (label is not null)
        {
            attributes["aria-label"] = label;
        }

        return attributes;
    }

    protected override Task OnComponentAfterRenderAsync(bool firstRender) =>
        JsModuleManager.InvokeModuleVoidAsync(
            ModuleName,
            "attach",
            InstanceId,
            RootElement,
            Tilt && !Disabled,
            MaxTiltAngle,
            HasSheen);

    private bool IsControlled => IsFlippedChanged.HasDelegate;

    private bool CurrentFlipped => IsControlled ? IsFlipped : _internalFlipped;

    private bool CanFlip => Back is not null;

    /// <summary>
    /// Click flip is the keyboard reachable control state: the card owns the
    /// pressed state and hides the face that is not showing.
    /// </summary>
    private bool UsesClickFlip => CanFlip && FlipTrigger == FlashCardFlipTrigger.Click;

    /// <summary>
    /// Hover flip is a pointer-only preview driven by the isolated stylesheet. It
    /// never changes state, so neither face is hidden from assistive technology and
    /// the card exposes no button semantics.
    /// </summary>
    private bool HoverPreview => CanFlip && FlipTrigger == FlashCardFlipTrigger.Hover;

    private bool IsBackVisible => UsesClickFlip && CurrentFlipped;

    private bool IsBackHidden => UsesClickFlip && !CurrentFlipped;

    private bool HasImage => !string.IsNullOrWhiteSpace(ImageSrc);

    /// <summary>
    /// The sheen only exists over media: a card without an image has no surface to
    /// reflect, and covering the caption would dim the text it carries.
    /// </summary>
    private bool HasSheen => HasImage && Sheen != FlashCardSheen.None;

    // Each mapper returns null for the tier that the component's base rule already
    // renders, so no modifier class is emitted without a matching rule.
    private string? SheenClass => Sheen switch
    {
        FlashCardSheen.Shine => "sheen-shine",
        FlashCardSheen.Holo => "sheen-holo",
        _ => null
    };

    private bool HasTitle => !string.IsNullOrWhiteSpace(Title);

    private bool HasDescription => !string.IsNullOrWhiteSpace(Description);

    private bool HasCaption => HasTitle || HasDescription;

    private string ImageAltValue => ImageAlt?.Trim() ?? string.Empty;

    /// <summary>
    /// Click handler bound only for flippable cards. A default
    /// <see cref="EventCallback{T}" /> renders no attribute at all, so a static
    /// card registers no DOM listener.
    /// </summary>
    private EventCallback<MouseEventArgs> ClickHandler => UsesClickFlip
        ? EventCallback.Factory.Create<MouseEventArgs>(this, HandleClickAsync)
        : default;

    /// <summary>Keyboard handler bound only for click flipped cards.</summary>
    private EventCallback<KeyboardEventArgs> KeyDownHandler => UsesClickFlip
        ? EventCallback.Factory.Create<KeyboardEventArgs>(this, HandleKeyDownAsync)
        : default;

    private async Task HandleClickAsync(MouseEventArgs _) => await ToggleAsync();

    private async Task HandleKeyDownAsync(KeyboardEventArgs args)
    {
        if (args.Key is not ("Enter" or " "))
        {
            return;
        }

        await ToggleAsync();
    }

    private async Task ToggleAsync()
    {
        if (!UsesClickFlip || Disabled)
        {
            return;
        }

        var next = !CurrentFlipped;
        if (!IsControlled)
        {
            _internalFlipped = next;
        }

        if (IsFlippedChanged.HasDelegate)
        {
            await IsFlippedChanged.InvokeAsync(next);
        }
    }

    private static string? FirstNonBlank(params string?[] values) =>
        values.FirstOrDefault(value => !string.IsNullOrWhiteSpace(value))?.Trim();
}
